util/merkle/src/snapshot/chunk.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2019 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::collections::VecDeque;
+use std::convert::From;
+
+use ccrypto::BLAKE_NULL_RLP;
+use cdb::{DBValue, HashDB, MemoryDB};
+use primitives::H256;
+
+use super::error::{ChunkError, Error};
+use super::{DecodedPathSlice, PathSlice, CHUNK_HEIGHT};
+use crate::nibbleslice::NibbleSlice;
+use crate::{Node, TrieDBMut};
+
+#[derive(RlpEncodable, RlpDecodable, Eq, PartialEq)]
+pub struct TerminalNode {
+    // Relative path from the chunk root.
+    pub path_slice: PathSlice,
+    pub node_rlp: Vec<u8>,
+}
+
+impl std::fmt::Debug for TerminalNode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let path_slice = NibbleSlice::from_encoded(&self.path_slice);
+        f.debug_struct("TerminalNode")
+            .field("path_slice", &path_slice)
+            .field("node_rlp", &NodeDebugAdaptor {
+                rlp: &self.node_rlp,
+            })
+            .finish()
+    }
+}
+
+struct NodeDebugAdaptor<'a> {
+    rlp: &'a [u8],
+}
+
+impl<'a> std::fmt::Debug for NodeDebugAdaptor<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        match Node::decoded(&self.rlp) {
+            Some(node) => write!(f, "{:?}", &node),
+            None => write!(f, "{:?}", self.rlp),
```

Comment:
When does it become None?
Is it normal that the rlp cannot be decoded?

Reply:
I'm not sure why this would happen.
It seems `decoded` returns `None` when the rlp is empty, but this should not be happening in the cases that I know of.

Reply:
@Remagpie is right.

Reply:
Okay. Let's leave it for this PR and try to remove it in the next commit.